### PR TITLE
Updating the JIT to support marshaling blittable generics.

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/coreclr/issues/1685

### CODE REVIEW

**Comment by tannergooding** [comment]
Still TODO:
[ ] Add tests covering array marshaling
[ ] Add tests covering Layout Class marshaling
[ ] Add tests covering COM interop
[ ] Add tests covering `Vector64<T>` marshaling
[ ] Add tests covering `Vector128<T>` marshaling
[ ] Add tests covering `Vector256<T>` marshaling
[ ] Add tests covering marshaling generic pointers
[ ] Add tests covering marshaling ref/out blittable generics
[ ] Add negative tests covering non-blittable generics
[ ] Add negative tests explicitly covering `bool` and `char`

File: tests/src/Interop/PInvoke/Generics/GenericsNative.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include <stdio.h>
+#include <stdint.h>
+#include <xplatform.h>
+#include <platformdefines.h>
+
+struct Point1F
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/Generics/GenericsNative.cpp]: I would like to see 64-bit integers as well as `double` types tested as well.

File: tests/src/Interop/PInvoke/Generics/GenericsTest.cs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    [DllImport(nameof(GenericsNative))]
+    public static extern Point4<uint> GetPoint4U();
+
+    [DllImport(nameof(GenericsNative))]
+    public static extern Point4<uint> AddPoint4U(Point4<uint> lhs, Point4<uint> rhs);
+}
+
+class GenericsTest
+{
+    public unsafe static int Main(string[] args)
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/Generics/GenericsTest.cs]: Is `unsafe` needed?

File: tests/src/Interop/PInvoke/Generics/GenericsTest.cs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        {
+            Console.WriteLine(ex);
+            return 0;
+        }
+        return 100;
+    }
+
+    private static void TestPoint1F()
+    {
+        GenericsNative.Point1<float> value = GenericsNative.GetPoint1F();
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/Generics/GenericsTest.cs]: I really appreciate the value being in the name of the function, but would prefer functions that accept a passed in value instead.  The native side should contain nothing that is 'expected' by the test.

Ideally tests should be written in a way where all data for validation is in the test itself rather than returned from external assemblies or components.

```
float v = 1.0f;
GenericsNative.Point1<float> value = GenericsNative.GetPoint1F(v);
Assert.AreEqual(value.x, v);
```

File: src/dlls/mscorrc/mscorrc.rc [hunk 4]
```diff
@@ -626,7 +626,7 @@ BEGIN
     IDS_EE_BADMARSHAL_NOTMARSHALABLE        "The type definition of this field has layout information but has an invalid managed/unmanaged type combination or is unmarshalable."
     IDS_EE_BADMARSHAL_BADMETADATA           "Invalid marshaling metadata."
     IDS_EE_BADMARSHAL_CUSTOMMARSHALER       "Custom marshalers are only allowed on classes, strings, arrays, and boxed value types."
-    IDS_EE_BADMARSHAL_GENERICS_RESTRICTION  "Generic types cannot be marshaled."
+    IDS_EE_BADMARSHAL_GENERICS_RESTRICTION  "Non-blittable generic types cannot be marshaled."
```
Comment (jkoritzinsky) [inline src/dlls/mscorrc/mscorrc.rc]: If we arent changing the generics restriction everywhere, we should have both error messages and use the new one where we allow blittable generics.

File: tests/src/Interop/PInvoke/Generics/GenericsTest.cs (hunk 3 above)
Comment (tannergooding) [reply tests/src/Interop/PInvoke/Generics/GenericsTest.cs]: I can certainly add some tests that do that; but I think it is just as important to test the real world scenario of: The result is created and returned by native code.

There is a lot of native code out there that is `Foo GetX` and `void GetX(Foo* pResult)`

File: tests/src/Interop/PInvoke/Generics/GenericsNative.cpp (hunk 1 above)
Comment (tannergooding) [reply tests/src/Interop/PInvoke/Generics/GenericsNative.cpp]: Can do.

File: tests/src/Interop/PInvoke/Generics/GenericsTest.cs (hunk 2 above)
Comment (tannergooding) [reply tests/src/Interop/PInvoke/Generics/GenericsTest.cs]: Not here, will fix.

File: tests/src/Interop/PInvoke/Generics/GenericsTest.cs (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply tests/src/Interop/PInvoke/Generics/GenericsTest.cs]: Agreed, but that scenario is covered by passing in a value as well. There is no need for test function that takes no arguments and one that does. At least not in this scenario.

Comment (tannergooding) [reply tests/src/Interop/PInvoke/Generics/GenericsTest.cs]: Fixed.

File: tests/src/Interop/PInvoke/Generics/GenericsNative.cpp (hunk 1 above)
Comment (tannergooding) [reply tests/src/Interop/PInvoke/Generics/GenericsNative.cpp]: Fixed.

**Comment by tannergooding** [comment]
Responded to feedback so far; added tests covering arrays, points, ref, and out.

I'm going to be camping until Monday, but I'll see how far I can get this tonight.

**Comment by jkoritzinsky** [comment]
Can you make sure to add a test for a non-generic struct containing a field that has a type of an instantiated generic struct?

**Comment by tannergooding** [comment]
CC. @CarolEidt. You might be interested in this from the perspective of the vector ABI support

**Comment by tannergooding** [comment]
For the `Vector64<T>`, `Vector128<T>`, and `Vector256<T>` code; we will need tests disabled on some platforms until the JIT ABI can be fixed. For example, the JIT for Windows x64 is not currently returning `__m128` in register; but it should be.

Will also need to figure out how to handle cases where a given platform doesn't quite support a given type. For example, ARM does not support `Vector128<double>`, but ARM64 does. This doesn't really matter from an ABI perspective (since all vectors are passed/returned the same), so it might be okay to skip one of the `T` for a given platform, provided we have a few other samples that do work...

**Comment by tannergooding** [comment]
Rebased onto current head.

**Comment by jkoritzinsky** [comment]
You might need to add an explicit block for `Span<T>` and `ReadOnlySpan<T>`. This change likely will make them seem blittable (which they shouldn't be).

**Comment by tannergooding** [comment]
> You might need to add an explicit block for Span<T> and ReadOnlySpan<T>

I'm guessing this is because `ByReference<T>` only contains an `IntPtr` internally, correct?

**Comment by jkoritzinsky** [comment]
Yep that's why. We actually probably just want to prohibit `ByReference<T>` since that would handle both cases.

**Comment by jkotas** [comment]
> For example, the JIT for Windows x64 is not currently returning __m128 in register; but it should be.

We should keep blocking Vector128 and similar cases for interop until this is fixed.

**Comment by tannergooding** [comment]
> We should keep blocking Vector128 and similar cases for interop until this is fixed.

I'll update to explicitly block these and `ByReference<T>` for the time being

**Comment by jkoritzinsky** [comment]
Can you also ensure that `Nullable<T>` is still blocked outside of WinRT scenarios so that we aren't locked into the default behavior in case we decide to do something special with the type?

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

**Comment by AaronRobinsonMSFT** [comment]
@tannergooding What are you thoughts for this PR with respect to repo consolidation? I am going to assume you would like this in prior to that happening. Please confirm so @jkoritzinsky and myself can prioritize reviewing this PR.

**Comment by tannergooding** [comment]
@AaronRobinsonMSFT, I'm not particularly concerned either way. I just had some time to finally add some of the tests that were missing.

If I happen to get all needed test coverage and CI passing, then I'm fine with it being reviewed and merged before the consolidation.

Otherwise, I'm also perfectly comfortable shuffling these changes over into a new PR 😄

**Comment by AaronRobinsonMSFT** [comment]
I looked at the product changes and they seem reasonable with no concerns from me. I want to peek a bit more at some of the tests themselves, but otherwise if the CI is green and coverage is what has been agreed upon I am signed off. Thanks for driving this work - much appreciated!

**Comment by tannergooding** [comment]
> coverage is what has been agreed upon I am signed off

I still need tests validating `Nullable<T>` is blocked for non-WinRT scenarios (and I believe that will require a product change as well).

Ideally, I would also have tests covering (validating they don't work) generic COM types, generic classes with `LayoutKind.Sequential`, and explicit tests covering `MyStruct<ReferenceType>`.

Right now I'm trying to validate I have the `#includes` right for `Vector64/128/256` on ARM/ARM64. Namely the headers for Windows vs Unix differed for ARM64 (just fixed) and I believe clang is complaining about NEON not being enabled for ARM32.

**Comment by tannergooding** [comment]
It looks like for ARM32 we target `-mfpu=vfpv3` (https://github.com/dotnet/coreclr/blob/master/configurecompiler.cmake#L542) and even support softfp as a fallback.

`arm_neon.h` therefore can't be included as this needs to be `-mfpu=neon` instead. Is there any particular reason we target `armv7-a` but don't also target `neon`? It was my understanding that, outside of some edge scenarios, armv7a implied neon support (maybe @TamarChristinaArm could comment on this)...

**Comment by tannergooding** [comment]
@CarolEidt, were you planning on having `Vector<T>` an ABI primitive (like `Vector128<T>`/`Vector256<T>` or a user-defined struct (like `Vector2`/`Vector3`/`Vector4`)?

Just want to make sure I add the appropriate tests here and disallow it for the time being if necessary 😄

**Comment by jkotas** [comment]
`Vector<T>` should be prohibited from interop. There is no way to use it for interop in reasonable way since it has unpredictable size.

**Comment by tannergooding** [comment]
@jkotas, what would be the appropriate way to handle `Vector<T>*`.

That is, as of C# 8 since `Vector<T>*` is now legal code (and prior to that in IL), so `public static extern void Method(Vector<int>* pValue)` is currently valid (even without this change).

Should that just be allowed and people are on their own or should we poison even pointers to `Vector<T>` from being marshalled (this would be a technically breaking change, but I doubt anyone would be broken in practice)?

In a somewhat similar vein, `Vector128<T>` is "valid" even for unsupported `T`. For example `Vector128<bool>` or even `Vector128<Vector256<int>>` is "fine". This is true even in managed code where `= default` and `= new Vector128<T>()` works, but all instance methods will throw due to an unsupported `T`.

**Comment by jkotas** [comment]
> we poison even pointers to Vector<T> from being marshalled

Unmanaged pointers to anything are fine with me.

Do we prohibit any unmanaged pointers in interop today? I would not mind removing any restrictions if there are any.

**Comment by tannergooding** [comment]
We prohibit `bool*` and pointers to anything containing a `bool`. It gives `Pointers cannot reference marshaled structures.  Use ByRef instead.'`

image

There may be others, but that is the one that immediately pops to mind.

**Comment by jkoritzinsky** [comment]
We prohibit unmanaged pointers to non-blittable types in parameters and return values.

We've never been able to do so to structs because of the fact that the checking happened in the type loader and we'd end up with a seemingly cyclical type. For example, in the type below:

```csharp
unsafe struct D
{
     D* ptr;
}
```

We couldn't validate if `D` was blittable since we were still loading `D`.

In the parameter and return value case, the types are fully loaded so we can do the validation.

If we want to allow marshalling unmanaged pointers to nonblittable "unmanaged" types as parameters and return values, I'd be ok with removing that restriction.

**Comment by TamarChristinaArm** [comment]
> `arm_neon.h` therefore can't be included as this needs to be `-mfpu=neon` instead. Is there any particular reason we target `armv7-a` but don't also target `neon`? It was my understanding that, outside of some edge scenarios, armv7a implied neon support (maybe @TamarChristinaArm could comment on this)...
> 

Architecturally, you can have `Armv7-A` without `NEON` or `VFP`, and any mix of the two options. Realistically though most hardware you can buy off the shelf when they have `NEON` have `VFP` as well...

> _Edit:_ Looks like GCC documents `-mfpu=neon` to be an alias for `-mfpu=neon-vfpv3` (https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html). LLVM doesn't look to explicitly document these switches so I would guess they are part of the switches that are compatible with GCC

I've asked the LLVM team and `neon` does at least enable `vfp3` there as well. The full set it enables is:

```
"-target-cpu" "generic"
"-target-feature" "+soft-float-abi"
"-target-feature" "+fpregs"
"-target-feature" "+vfp2"
"-target-feature" "+vfp2sp"
"-target-feature" "+vfp3"
"-target-feature" "+vfp3d16"
"-target-feature" "+vfp3d16sp"
"-target-feature" "+vfp3sp"
"-target-feature" "+fp64"
"-target-feature" "+d32"
"-target-feature" "+neon"
"-target-feature" "+strict-align"
```

**Comment by CarolEidt** [comment]
I agree with @jkotas that it makes no sense to allow `Vector<T>` to be marshalled. I don't have a strong opinion on whether one should allow `Vector<T>*`, though I would lean toward disallowing it, as it also seems a bit nonsensical.
It would seem that if one were doing interop with vector types, one would be better off with the fixed-size HW intrinsic types, and if one wants to interop with `Vector<T>` one should convert to/from fixed-size HW intrinsic types.

**Comment by tannergooding** [comment]
I've disallowed `Vector<T>` and written some tests that correctly handle the `Vector<T>*` case for the time being. It's technically a breaking change to disallow it now (although likely a very small one) and we would likely need some more extensive analysis to disallow it (we would need to track if anything in the pointer type, even transitively, was `Vector<T>`).

**Comment by tannergooding** [comment]
Not necessarily as part of this PR, but as part of the ABI work (https://github.com/dotnet/coreclr/issues/15943), we'll need to determine how to handle `Vector64<T>`, `Vector128<T>`, and `Vector256<T>` when they are "unsupported". That is:
* If the JIT is compiled with `FEATURE_SIMD` off, what is the behavior?
* If the user is running on hardware without AVX support, should `Vector256<T>` be disallowed?
* If the user explicitly disables VEX support (`COMPlus_EnableAVX=0`) should `Vector256<T>` be disallowed?
* For `Vector64<T>`, where it technically tracks `__m64` on x86; but where we never plan on exposing `__m64` (due to it being legacy) should `Vector64<T>` be disallowed?

In all cases, my initial preference would be to disallow the scenario from working. That prevents users from taking a dependency on these being treated like "user-defined" structs in some scenarios and leaves it open for us to support later if needed.

**Comment by TamarChristinaArm** [comment]
> It would seem that if one were doing interop with vector types, one would be better off with the fixed-size HW intrinsic types, and if one wants to interop with Vector<T> one should convert to/from fixed-size HW intrinsic types.

Sorry, haven't been following the discussion much, but wouldn't this cause a problem in the future with something like SVE? where the user really doesn't know the size of the Vector?

It's probably not relevant for now but just wondering.

**Comment by tannergooding** [comment]
Please correct me if I'm wrong, but my understanding of SVE is that there are a number of restrictions on them including treating them as incomplete types in higher level languages (and not allowing `sizeof` or use in arrays, etc).
Additionally, the size can be changed at execution time by modifying `ZCR_ELx.LEN`
However, the reciever/return value is always going to be the "sizeless" type and the instructions themsleves are always the same, regardless of the "actual size"; is that correct?

I believe that differs from x86 (and therefore `Vector<T>`, it being cross-platform) in that on x86 you have different instructions, different types, and even different registers (even if some are subsets of others) for `Vector128<T>` vs `Vector256<T>`.

If/when we expose SVE, I would imagine it would need to be an ARM specific type (under `S.R.I.Arm`) rather than a cross-platform type (under `S.R.Intrinsics`, like `Vector64<T>`, `Vector128<T>`, and `Vector256<T>`) and there would need to be very specialized handling to ensure it always goes through the appropriate instructions for access, etc.

**Comment by CarolEidt** [comment]
I don't see a strong reason to disallow the fixed-size types when on a target that doesn't support them. Doing so would require users, in scenarios where they want to provide backing code for the non-accelerated case, to wrap or otherwise obfuscate these types unnecessarily.

> If the JIT is compiled with FEATURE_SIMD off, what is the behavior?

I think it makes sense for the fixed-size types to be recognized and supported, but not any intrinsics defined on or using those types.

> If the user is running on hardware without AVX support, should Vector256<T> be disallowed?
If the user explicitly disables VEX support (COMPlus_EnableAVX=0) should Vector256<T> be disallowed? For Vector64<T>, where it technically tracks __m64 on x86; but where we never plan on exposing __m64 (due to it being legacy) should Vector64<T> be disallowed?

I don't see a benefit to disallowing these.

> That prevents users from taking a dependency on these being treated like "user-defined" structs in some scenarios and leaves it open for us to support later if needed.

Can you suggest the kind of scenario where this would cause an issue? My imagination is failing me here.

**Comment by tannergooding** [comment]
@CarolEidt, the scenario is that the JIT doesn't support the target types but the native code does.

For example, `Vector256<T>` is meant to track the ABI primitive `__m256`. The user has disabled AVX support, so the JIT views `Vector256<T>` as a "user-defined" type in managed code. However, the expectation when passing it to native code is that it is treated as `__m256`. It is now marshaled incorrectly (as a user-defined type) and results in corrupted data.

Restricting basically says: "this is always the ABI primitive type, never a 'user-defined' struct". So, the user can always expect that it is correctly handled as the ABI primitive (`__m64`, `__m128`, `__m256`, and corresponding types on ARM) and that it will fail if the JIT (for whatever reason) doesn't support those types.

**Comment by tannergooding** [comment]
It might also be worth noting that the proposal is to only disallow for the interop case; creating and using these types entirely in managed code would still be fine. The proposal is to only avoid errors when transitioning outside of managed code.

**Comment by CarolEidt** [comment]
> the scenario is that the JIT doesn't support the target types but the native code does.

Right. So perhaps it makes sense to disallow types that cannot be supported (i.e. I can't appropriately pass `Vector256<T>` to native code if I'm not using VEX encodings), or for which there is no defined ABI behavior.
- On x86 we can always pass both `Vector128<T>` and `Vector64<T>` correctly even if there are no intrinsics defined/available on those types.
- Where there is (as yet) no specific ABI defined for a type - e.g. `Vector256<T>` on Arm architectures - I would say that we disallow the type.

For Arm32, the types should be disallowed until we support whatever the ABI requirements are for that target - but it seems to me that we could do that independently of adding support for the actual intrinsics.

**Comment by tannergooding** [comment]
> On x86 we can always pass both Vector128<T> and Vector64<T> correctly even if there are no intrinsics defined/available on those types.

Is that true even when `FEATURE_SIMD` is disabled? If so, that's great and helps things 😄 

However, it's worth noting we can't currently pass `Vector64<T>` correctly. It technically corresponds to `__m64` and by value returns (on 32-bit Windows at least) needs to be passed in `mm0`. We have no support for in the JIT today for these MMX registers as MMX is considered legacy and is effectively deprecated.

**Comment by CarolEidt** [comment]
>> On x86 we can always pass both Vector128 and Vector64 correctly even if there are no intrinsics defined/available on those types.

> Is that true even when FEATURE_SIMD is disabled? If so, that's great and helps things 😄

It is not currently true, but I think that we should make it so. What I meant by "We can always ..." was "We should be able to always ..."

> However, it's worth noting we can't currently pass Vector64<T> correctly. It technically corresponds to __m64 and by value returns (on 32-bit Windows at least) needs to be passed in mm0. We have no support for in the JIT today for these MMX registers as MMX is considered legacy and is effectively deprecated.

Right - I'd forgotten that the MMX registers are aliased to the FP stack registers :-(  Although we could theoretically support that (we currently have to return float values on the stack for x86, it probably isn't worth it.

**Comment by tannergooding** [comment]
There looks to be some kind of bug in `Unsafe.AddByteOffset` and `Unsafe.Add`:
image

**Comment by tannergooding** [comment]
I forgot to clear the MMX state for the Vector64<T> native code (which uses `__m64`), it's fixed locally.

**Comment by tannergooding** [comment]
Ok, I should just need to add explicit tests covering classes with `LayoutKind.Sequential` and generic COM types now and then all requested scenarios should be covered.

I'm going to log an issue for enabling unmanaged pointers to non-blittable types (https://github.com/dotnet/coreclr/pull/23899#issuecomment-551374665) and follow up with it in a subsequent PR.

**Comment by tannergooding** [comment]
CC. @AaronRobinsonMSFT, @jkoritzinsky, @jkotas.

I believe this should be ready for review now. I've done the following:
* Logged a bug tracking enablement of unmanaged pointers to non-blittable types (such as `bool`): https://github.com/dotnet/coreclr/issues/27800
* Updated src/vm/mlinfo.cpp to allow blittable generics
  * ByReference<T>, Nullable<T>, and Vector<T> are explicitly blocked with no plans for unblocking
  * Vector64<T>, Vector128<T>, and Vector256<T> are currently blocked pending https://github.com/dotnet/coreclr/issues/15943
* Added explicit tests covering various scenarios (for both allowed and disallowed cases)
  * The scenarios should be summed up here: https://github.com/dotnet/coreclr/pull/23899#issuecomment-482219021

**Comment by tannergooding** [comment]
The vast majority of changes here are tests covering the same set of scenarios:
* ref/out/in/pointer, by value, and array parameters
* ref/ref readonly returns

The types covered by the tests then touch:
* float/double, uint/long, and bool/char

These validate that various important ABI conventions are being correctly handled (such as HFA/HVA, two field structs that can be passed in register, etc).

File: src/jit/lowerxarch.cpp [hunk 5]
```diff
@@ -2901,6 +2901,12 @@ bool Lowering::IsContainableHWIntrinsicOp(GenTreeHWIntrinsic* containingNode, Ge
             return supportsUnalignedSIMDLoads;
         }
 
+        case NI_AVX_ExtractVector128:
```
Comment (jkotas) [inline src/jit/lowerxarch.cpp]: Is this independent JIT bug fix? It may be best to submit this JIT bug fix as separate PR, so that it gets the right level of scrutiny as other JIT bug fixes.

File: src/vm/mlinfo.cpp [hunk 6]
```diff
@@ -2796,9 +2796,18 @@ MarshalInfo::MarshalInfo(Module* pModule,
                 }
 #endif // FEATURE_COMINTEROP
 
-                if (m_pMT->HasInstantiation())
+                if (m_pMT->HasInstantiation() && (!m_pMT->IsBlittable()
+                    || m_pMT->HasSameTypeDefAs(g_pByReferenceClass)
+                    || m_pMT->HasSameTypeDefAs(g_pNullableClass)
```
Comment (jkotas) [inline src/vm/mlinfo.cpp]: What is special about nullable to block it? Maybe worth a comment.

File: src/vm/mlinfo.cpp [hunk 7]
```diff
@@ -2796,9 +2796,18 @@ MarshalInfo::MarshalInfo(Module* pModule,
...
+                    || m_pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__VECTOR64T))
+                    || m_pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__VECTOR128T))
+                    || m_pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__VECTOR256T))
+#ifndef CROSSGEN_COMPILE
```
Comment (jkotas) [inline src/vm/mlinfo.cpp]: Why is this ifndef necessary? We try to avoid logic differences like this for crossgen.

File: src/dlls/mscorrc/mscorrc.rc [hunk 8]
```diff
@@ -629,6 +629,7 @@ BEGIN
     IDS_EE_BADMARSHAL_CUSTOMMARSHALER       "Custom marshalers are only allowed on classes, strings, arrays, and boxed value types."
     IDS_EE_BADMARSHAL_GENERICS_RESTRICTION  "Generic types cannot be marshaled."
     IDS_EE_BADMARSHAL_STRING_OUT            "Cannot marshal a string by-value with the [Out] attribute."
+    IDS_EE_BADMARSHAL_BLITTABLE_GENERICS_RESTRICTION "Non-blittable generic types cannot be marshaled."
```
Comment (jkotas) [inline src/dlls/mscorrc/mscorrc.rc]: Is the `IDS_EE_BADMARSHAL_GENERICS_RESTRICTION` string still relevant? Would it be more accurate to replace all instances of the old string with the new string?

File: src/vm/mlinfo.cpp (hunk 6 above)
Comment (tannergooding) [reply src/vm/mlinfo.cpp]: @jkoritzinsky asked me to here: https://github.com/dotnet/coreclr/pull/23899#issuecomment-538528286

Its to avoid locking us into the default behavior in case we decide to do something special with the type.

File: src/jit/lowerxarch.cpp (hunk 5 above)
Comment (tannergooding) [reply src/jit/lowerxarch.cpp]: Yes, it's just avoiding the assert from the default case.

Will extract.

File: src/vm/mlinfo.cpp (hunk 7 above)
Comment (tannergooding) [reply src/vm/mlinfo.cpp]: `Vector<T>` is blocked from even being type loaded during crossgen (due to its variable size): https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L1182

File: src/dlls/mscorrc/mscorrc.rc (hunk 8 above)
Comment (tannergooding) [reply src/dlls/mscorrc/mscorrc.rc]: This is still being used when marshalling generic classes (modulo WinRT interfaces):
* https://github.com/dotnet/coreclr/blob/master/src/vm/mlinfo.cpp#L2022
* https://github.com/dotnet/coreclr/blob/master/src/vm/mlinfo.cpp#L2116

If there are no concerns replacing these strings (I don't see anything obvious), then I'll get a fix up.

File: src/jit/lowerxarch.cpp (hunk 5 above)
Comment (tannergooding) [reply src/jit/lowerxarch.cpp]: Extracted to https://github.com/dotnet/coreclr/pull/27807

File: src/vm/mlinfo.cpp (hunk 7 above)
Comment (jkotas) [reply src/vm/mlinfo.cpp]: Ok, make sense. I have not realized that this is to avoid failing here 100% of time during crossgen. Comment may be nice

Comment (jkotas) [reply src/vm/mlinfo.cpp]: Is there a matching change to be done for this in crossgen2?

Comment (tannergooding) [reply src/vm/mlinfo.cpp]: I'm not sure. Looking at crossgen2, I don't see any specific handling of `Vector<T>` and disallowing it as a type.

I did find a section for the FieldLayoutAlgorithm: https://github.com/dotnet/coreclr/blob/master/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilerContext.cs#L50
However, that field layout only supports `Vector64/128/256<T>`: https://github.com/dotnet/coreclr/blob/master/src/tools/crossgen2/Common/Compiler/VectorFieldLayoutAlgorithm.cs

The only logic I can find for `Vector<T>` is in the `SystemVStructClassificator` where it doesn't treat it as an EightBytes struct: https://github.com/dotnet/coreclr/blob/master/src/tools/crossgen2/Common/JitInterface/SystemVStructClassificator.cs#L273

Comment (jkotas) [reply src/vm/mlinfo.cpp]: We may need a change here: https://github.com/dotnet/coreclr/blob/master/src/tools/crossgen2/Common/TypeSystem/Interop/IL/MarshalHelpers.cs#L337

cc @fadimounir

**Comment by TamarChristinaArm** [comment]
> Please correct me if I'm wrong, but my understanding of SVE is that there are a number of restrictions on them including treating them as incomplete types in higher level languages (and not allowing sizeof or use in arrays, etc).

That's correct, the SVE ACLE types we have are incomplete, sizeless and definite.

> Additionally, the size can be changed at execution time by modifying ZCR_ELx.LEN

Not by anything user-mode. While technically the kernel is allowed to change it, the expectation is that it won't do this after the process has started.  You can have processes with different VLs on the same system though.

> However, the reciever/return value is always going to be the "sizeless" type and the instructions themsleves are always the same, regardless of the "actual size"; is that correct?

Yeah, "sizeless" of a specific type. So the way I saw it was that `Vector<T>` would represent all the SVE sizeless types, e.g. `svint8_t` would be `Vector<Int8>`. As in, you don't know how large the vector is but you do know it's element size.

SVE also allows compiling code for a specific VL, in which case in ACLE you are then allowed to cast between the incomplete and a known complete type.

> I believe that differs from x86 (and therefore Vector<T>, it being cross-platform) in that on x86 you have different instructions, different types, and even different registers (even if some are subsets of others) for Vector128<T> vs Vector256<T>.

Ah, wait, I think I'm missing something here.. Is `Vector<T>` not an actual type but a name for the  "grouping" of all `Vector*<T>` types? But yes for SVE you wouldn't have different function or registers for different VLs.

> If/when we expose SVE, I would imagine it would need to be an ARM specific type (under S.R.I.Arm) rather than a cross-platform type (under S.R.Intrinsics, like Vector64<T>, Vector128<T>, and Vector256<T>) and there would need to be very specialized handling to ensure it always goes through the appropriate instructions for access, etc.

Hmm perhaps... that said there are other ISAs other than SVE which have the same understanding as VL agnostic types.  So I wouldn't necessarily say this type would need to be SVE specific.

**Comment by tannergooding** [comment]
@TamarChristinaArm, I'ved forked the discussion off to: https://github.com/dotnet/coreclr/issues/27814

File: src/vm/mlinfo.cpp (hunk 6 above)
Comment (tannergooding) [reply src/vm/mlinfo.cpp]: Added a comment

File: src/dlls/mscorrc/mscorrc.rc (hunk 8 above)
Comment (tannergooding) [reply src/dlls/mscorrc/mscorrc.rc]: Fixed.

File: src/vm/mlinfo.cpp (hunk 7 above)
Comment (tannergooding) [reply src/vm/mlinfo.cpp]: What is the easiest way to test Crossgen2 on an specific test?

Comment (tannergooding) [reply src/vm/mlinfo.cpp]: I've pushed up a commit which I believe handles this, but it's not been tested (and doesn't look to be covered by CI currently).

**Comment by tannergooding** [comment]
Closing as this has been ported to dotnet/runtime repo.
